### PR TITLE
more python2/3 compat changes (SOFTWARE-4348)

### PR DESCRIPTION
**Pull request by edquist**
Right, so, sadly we have to convert all this stuff - at least the stuff with broken python3 syntax - for the el7 build with python3.

On the bright side (?), hopefully most of this isn't used anymore anyway...

### CODE REVIEW

File: onevm/gratia/onevm/process_utils.py [hunk 1]
```diff
@@ -1,12 +1,4 @@
-import popen2
-try:
-    import cStringIO
-except ImportError:
-    import io as cStringIO
-import fcntl
-import os
-import time
-import select
+from subproces import Popen, PIPE
```
Comment (matyasselmeci) [inline onevm/gratia/onevm/process_utils.py:1]:
```suggestion
from subprocess import Popen, PIPE
```

File: onevm/gratia/onevm/process_utils.py [hunk 2]
```diff
@@ -34,51 +25,12 @@ def iexe_cmd(cmd, stdin_data=None):
     error_lines = None
     exit_status = 0
     try:
-        # XXX: no popen2 in python3
-        child = popen2.Popen3(cmd, capturestderr=True)
+        p = Popen(cmd, shell=True, stdin=PIPE, stdout=PIPE, stderr=PIPE)
+        out, err = p.communicate(stdin_data)
```
Comment (matyasselmeci) [inline onevm/gratia/onevm/process_utils.py:29]: Careful: communicate() takes and returns bytes here.

Comment (edquist) [reply onevm/gratia/onevm/process_utils.py:30]: Right...

Comment (edquist) [reply onevm/gratia/onevm/process_utils.py:30]: I am going to wimp out and use universal_newlines here

File: xrootd-storage/gratia-xrootd-storage [hunk 3]
```diff
@@ -97,7 +100,7 @@ class XrootdSummaryParser(ContentHandler):
     def emitData(self, key, val):
         try:
             self.dataCallback(key, val)
-        except SystemExit, KeyboardInterrupt:
+        except SystemExit as KeyboardInterrupt:
```
Comment (matyasselmeci) [inline xrootd-storage/gratia-xrootd-storage:103]:
```suggestion
        except (SystemExit, KeyboardInterrupt) as _:
```
I think this what you _want_ to do -- catch both SystemExit and KeyboardInterrupt; in this case, the two need to be in a tuple in Python 2. (What the previous code was doing in Python 2 was catching SystemExit and storing the exception in KeyboardInterrupt.)

File: xrootd-storage/gratia-xrootd-storage [hunk 4]
```diff
@@ -645,28 +648,28 @@ def main():
...
-	        time.sleep(SLEEP_TIME)
-	    except KeyboardInterrupt, SystemExit:
-	        raise
-	    except Exception, e:
-	        gratia_log_traceback(lvl=1)
+        while 1 == 1:
+            try:
+                DebugPrint(0, "Will send new Gratia data in %i seconds." % SLEEP_TIME)
+                time.sleep(SLEEP_TIME)
+            except KeyboardInterrupt as SystemExit:
```
Comment (matyasselmeci) [inline xrootd-storage/gratia-xrootd-storage:655]:
```suggestion
            except (KeyboardInterrupt, SystemExit) as _:
```

File: xrootd-storage/gratia-xrootd-storage [hunk 5]
```diff
@@ -645,28 +648,28 @@ def main():
...
-	        gratia_log_traceback(lvl=0)
-	        gratia_handler.stop_exception()
-	        gratia_handler.stop = True
-	        raise
-	    except Exception, e:
-	        gratia_log_traceback(lvl=1)
+            try:
+                DebugPrint(1, "Creating a new Xrootd Gratia report.")
+                gratia_handler.summary()
+            except KeyboardInterrupt as SystemExit:
```
Comment (matyasselmeci) [inline xrootd-storage/gratia-xrootd-storage:666]:
```suggestion
            except (KeyboardInterrupt, SystemExit) as _:
```

File: xrootd-transfer/gratia-xrootd-transfer [hunk 6]
```diff
@@ -769,21 +768,21 @@ def main():
...
-	        gratia_handler.summary()
-	    except KeyboardInterrupt, SystemExit:
-	        raise
-	    except Exception, e:
-	        gratia_log_traceback(lvl=1)
+        while 1 == 1:
+            try:
+                DebugPrint(0, "Will send new Gratia data in %i seconds." % SLEEP_TIME)
+                time.sleep(SLEEP_TIME)
+            except KeyboardInterrupt as SystemExit:
```
Comment (matyasselmeci) [inline xrootd-transfer/gratia-xrootd-transfer:775]:
```suggestion
            except (KeyboardInterrupt, SystemExit) as _:
```

File: xrootd-transfer/gratia-xrootd-transfer [hunk 7]
```diff
@@ -769,21 +768,21 @@ def main():
...
+                raise
+            except Exception as e:
+                gratia_log_traceback(lvl=1)
+            try:
+                DebugPrint(1, "Creating a new Xrootd Gratia report.")
+                gratia_handler.summary()
+            except KeyboardInterrupt as SystemExit:
```
Comment (matyasselmeci) [inline xrootd-transfer/gratia-xrootd-transfer:782]:
```suggestion
            except (KeyboardInterrupt, SystemExit) as _:
```

File: xrootd-transfer/gratia-xrootd-transfer-alt [hunk 8]
```diff
@@ -542,23 +541,23 @@ def main():
...
-	    except KeyboardInterrupt, SystemExit:
-	        raise
-	    except Exception, e:
-	        gratia_log_traceback(lvl=1)
-            except GratiaTimeout, gt:
+        while 1 == 1:
+            try:
+                DebugPrint(0, "Will send new Gratia data in %i seconds." % SLEEP_TIME)
+                time.sleep(SLEEP_TIME)
+            except KeyboardInterrupt as SystemExit:
```
Comment (matyasselmeci) [inline xrootd-transfer/gratia-xrootd-transfer-alt:548]:
```suggestion
            except (KeyboardInterrupt, SystemExit) as _:
```

File: xrootd-transfer/gratia-xrootd-transfer-alt [hunk 9]
```diff
@@ -542,23 +541,23 @@ def main():
...
+                raise
+            except Exception as e:
+                gratia_log_traceback(lvl=1)
+            except GratiaTimeout as gt:
+                gratia_log_traceback(lvl=1)
+            try:
+                DebugPrint(1, "Creating a new Xrootd Gratia report.")
+                gratia_handler.summary()
+            except KeyboardInterrupt as SystemExit:
```
Comment (matyasselmeci) [inline xrootd-transfer/gratia-xrootd-transfer-alt:557]:
```suggestion
            except (KeyboardInterrupt, SystemExit) as _:
```

**Review by matyasselmeci — CHANGES_REQUESTED** [review]
For the most part good but there's a few minor things that need fixing first.

File: condor-events/watchCondorEvents [hunk 10]
```diff
@@ -145,10 +146,10 @@ class ClassAdHandler(ContentHandler):
         if cur.get('MyType', 'UNKNOWN') == 'JobAdInformationEvent':
             if prev.get('MyType', 'UNKNOWN') == 'JobAdInformationEvent':
                 # Two info events in a row
-                print >> sys.stderr, "Two JobAdInformationEvents in a row!"
+                print("Two JobAdInformationEvents in a row!", file=sys.stderr)
                 return
             # A real event followed by an info event; combine and emit one.
-            for key, val in cur.items():
+            for key, val in list(cur.items()):
```
[the remark this answers is not in the document]
Comment (edquist) [reply condor-events/watchCondorEvents:152]: Man, i manually un-did like a hundred of these unnecessary list things that 2to3 inserted.  Missed this one i guess.

File: dCache-transfer/gratia/dcache_transfer/Checkpoint.py [hunk 11]
```diff
@@ -43,13 +45,18 @@ def __init__(self, tablename, maxAge=30):
...
             pklFile = open(tablename, 'rb')
-            self._dateStamp, self._transaction = cPickle.load(pklFile)
+            self._dateStamp, self._transaction = pickle.load(pklFile)
             if self._dateStamp < minDay:
                 self._dateStamp = minDay
             ds = self._dateStamp
             self._dateStamp = datetime(ds.year, ds.month, ds.day, ds.hour, 0, 0)
             pklFile.close()
-        except IOError, (errno, strerror):
+        except IOError as xxx_todo_changeme1:
```
[the remark this answers is not in the document]
Comment (edquist) [reply dCache-transfer/gratia/dcache_transfer/Checkpoint.py:54]: Errrgh.  I also manually fixed a bunch of these.........

[the remark this answers is not in the document]
Comment (edquist) [reply dCache-transfer/gratia/dcache_transfer/Checkpoint.py]: Fixing up

File: dCache-transfer/gratia/dcache_transfer/TimeBinRange.py [hunk 12]
```diff
@@ -110,7 +110,7 @@ def list(self):
         Returns a list of the aggregated rows held by this object.
         """
         result = []
-        tms =  self.bins.keys()
+        tms =  list(self.bins.keys())
         tms.sort()
         for tm in tms:
```
[the remark this answers is not in the document]
Comment (edquist) [reply dCache-transfer/gratia/dcache_transfer/TimeBinRange.py:115]: Yeah, that is better.  I was not making it my goal to improve code; just to make it work without making it much worse :-p

File: lsf/dev/filetest.py [hunk 13]
```diff
@@ -3,40 +3,41 @@
...
 
-print "** Tests setup -t: %s" % time.time()
+print("** Tests setup -t: %s" % time.time())
 sys.stdout.flush()
 
 #flength = 7
-import commands
-res = commands.getoutput('wc %s' % fname).split()
+import subprocess
+res = subprocess.getoutput('wc %s' % fname).split()
```
[the remark this answers is not in the document]
Comment (edquist) [reply lsf/dev/filetest.py]: Good catch.

File: lsf/dev/filetest.py [hunk 14]
```diff
@@ -211,9 +212,9 @@ def opcount(fname):
             assert func(fname) == flength
             counts[func].append(time.time() - start_time)
 
-    for key, vals in counts.items():
-        print key.__name__, ":", sum(vals) / float(len(vals))
+    for key, vals in list(counts.items()):
```
[the remark this answers is not in the document]
Comment (edquist) [reply lsf/dev/filetest.py:218]: Thanks; looks like i let this one slide also.

File: xrootd-storage/gratia-xrootd-storage (hunk 5 above)
Comment (edquist) [reply xrootd-storage/gratia-xrootd-storage:666]: Oh my, 2to3 messed that up, it seems...

**Comment by edquist** [comment]
Thanks Mat for the thorough review.  Hopefully all fixed now.

File: xrootd-storage/gratia-xrootd-storage (hunk 5 above)
Comment (matyasselmeci) [reply xrootd-storage/gratia-xrootd-storage]: Nope, this time it wasn't 2to3 -- that's what the original code was doing, it just wasn't obvious. That's why they added the _as_ syntax I guess.

**Review by matyasselmeci — APPROVED** [review]

**Comment by matyasselmeci** [comment]
Yep, looks good to me.

**Pull request merged**